src/cellrank/kernels/_base_kernel.py

```diff
@@ -463,6 +464,88 @@ def _reuse_cache(self, expected_params: Dict[str, Any], *, time: Optional[Any] =
             self._params = expected_params
         # fmt: on
 
+    def _get_mask(self, series, subset):
+        if isinstance(subset, str):
+            subset = [subset]
+        return series.isin(subset)
+
+    def get_boundary(self, source: str, target: str, cluster_key: str, graph: str = "distances"):
+        """Identify source observations at boundary to target cluster."""
+        source_obs_mask = self._get_mask(series=self.adata.obs[cluster_key], subset=source)
+        target_obs_mask = self._get_mask(series=self.adata.obs[cluster_key], subset=target)
+        obs_ids = np.arange(0, self.adata.n_obs)
+
+        source_ids = obs_ids[source_obs_mask]
+        adj_mat_subset = self.adata.obsp[graph][source_ids, :]
+
+        boundary_ids = []
+        for row_id, row in enumerate(adj_mat_subset):
+            _obs_mask = row.A.squeeze().astype(bool)
+            _obs_mask = _obs_mask & target_obs_mask
+
+            if _obs_mask.any():
+                boundary_ids.append(source_ids[row_id])
+
+        return boundary_ids
+
+    def get_empirical_velocity_field(self, boundary_ids, target_obs_mask, rep: str, graph: str = "distances"):
+        """Compute an emprical estimate of velocity field between two clusters."""
+        obs_ids = np.arange(0, self.adata.n_obs)
+        empirical_velo = pd.DataFrame(
+            index=obs_ids[boundary_ids], columns=[f"dim_{dim}" for dim in range(self.adata.obsm[rep].shape[1])]
+        )
+
+        adj_mat = self.adata.obsp[graph][boundary_ids, :]
+        for row_id, row in enumerate(adj_mat):
+            source_id = boundary_ids[row_id]
+            obs_mask = row.astype(bool).A.squeeze().copy() & target_obs_mask
+            neighbors = obs_ids[obs_mask]
+            weights = row.A.squeeze()[obs_mask]
+
+            empirical_velo.loc[source_id, :] = np.sum(
+                weights.reshape(-1, 1) * (self.adata[neighbors, :].obsm[rep] - self.adata[source_id, :].obsm[rep]),
+                axis=0,
+            )
+        empirical_velo = empirical_velo.dropna(axis=0).astype(float)
+
+        return empirical_velo.values
+
+    def get_vector_field_estimate(self, rep: str):
+        """Compute estimate of vector field under one step of the transition matrix."""
+        extrapolated_gex = self.transition_matrix @ self.adata.obsm[rep]
+        return extrapolated_gex - self.adata.obsm[rep]
+
+    def get_cbc(self, source: str, target: str, cluster_key: str, rep: str, graph: str = "distances"):
+        """Compute cross-boundary correctness score between source and target cluster."""
+
+        def get_pearson_corr(x, y):
```

Comment:
Can we use the Pearson correlation already implemented in CellRank, @michalk8?

Reply:
The one-dimensional case is redundant and an artifact; I refactored the function a bit to clean it up.
I tried using `_mat_mat_corr_dense` but the values are different - also compared to `scipy.stats.pearsonr`. Do you know why? I have a feeling it might be related to the degrees of freedom.

Reply:
Most likely yes, need to find the formulate exactly. Also `_mat_mat_corr_dense` computes **ALL** pairwise correlations, which is not what you need, so it shouldn't be used anyways.

Reply:
Yes, I know. I was going to subset to the diagonal. The current implementation works and is fast, though, so I'll just keep it.